File: legacy/app-desktop/src/model/Quiz.java
package model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Quiz {
	private List<Question> quiz;
	private List<Answer> answers; // user answers
	private int givenAnswers;
	private int correctAnswers;
		
	public Quiz(List<Question> questions, int qNum, boolean shuffleAnswers)
	{
		this.resetQuiz(questions, qNum, shuffleAnswers);
	}
	
	public Question getQuestionAt(int index) {return this.quiz.get(index);}
	public List<Question> getQuiz() {return this.quiz;}
	public List<Answer> getAnswers() {return this.answers;}
	public int getGivenAnswers() {return givenAnswers;}
	public int getCorrectAnswers() {return correctAnswers;}
	
	public void setAnswer(int index, int value)
	{
		/*if(value < 0 || value > Settings.ANSWER_NUMBER - 1)
			throw new IllegalArgumentException();*/
		if(this.answers.get(index).equals(Answer.NONE))
			this.givenAnswers++;
		
		this.answers.set(index, Answer.values()[value]);
	}
	public void setAnswer(int index, Answer value)
	{
		/*if(value < 0 || value > Settings.ANSWER_NUMBER - 1)
			throw new IllegalArgumentException();*/
		if(this.answers.get(index).equals(Answer.NONE))
			this.givenAnswers++;
		
		this.answers.set(index, value);
	}
	
	public void resetQuiz(List<Question> questions, int qNum, boolean shuffleAnswers)
	{
		// reset questions & user answers
		this.quiz = new ArrayList<Question>(qNum);
		this.answers = new ArrayList<Answer>(qNum);
		
		Collections.shuffle(questions);
		for(int i = 0; i < qNum; i++)
		{
			Question q = questions.get(i);
			
			if(shuffleAnswers)
				q.shuffleAnswers();
			
			this.quiz.add(q);
			this.answers.add(Answer.NONE);
		}
		
		// reset counters
		this.givenAnswers = 0;
		this.correctAnswers = 0;
	}
	
	public void checkAnswers()
	{
		for(int i = 0; i < this.quiz.size(); i++)
		{
			Answer ca = this.quiz.get(i).getCorrectAnswer();
			Answer ua = this.answers.get(i);
			
			if(ca.equals(ua))
				this.correctAnswers++;
		}
	}
}

File: legacy/app-mobile/flutter_application/lib/model/Quiz.dart
import 'package:roquiz/model/Answer.dart';
import 'package:roquiz/model/Question.dart';

class Quiz {
  List<Question> questions = [];
  List<Answer> answers = []; // user answers (selected or unselected)
  int givenAnswers = 0; // total given answers
  int correctAnswer = 0; // total correct answers

  /*Quiz(List<Question> questions, int qNum, bool shuffleAnswers) {
    _resetQuiz(questions, qNum, shuffleAnswers);
  }*/

  void resetQuiz(List<Question> questions, int qNum, bool shuffleAnswers) {
    this.questions = [];
    answers = [];

    // reset Questions
    questions.shuffle();
    for (int i = 0; i < qNum; i++) {
      if (shuffleAnswers) questions[i].shuffleAnswers();

      this.questions.add(questions[i]);
      answers.add(Answer.NONE);
    }

    // reset Answer counters
    givenAnswers = 0;
    correctAnswer = 0;
  }

  void setAnswerFromInt(int index, int value) {
    if (answers[index] == Answer.NONE) {
      givenAnswers++;
    }

    answers[index] = Answer.values[value];
  }

  void setAnswerFromEnum(int index, Answer value) {
    if (answers[index] == Answer.NONE) {
      givenAnswers++;
    }
    answers[index] = value;
  }

}
